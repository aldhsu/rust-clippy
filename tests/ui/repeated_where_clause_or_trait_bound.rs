// run-rustfix
//
#![allow(unused)]
#![deny(clippy::repeated_where_clause_or_trait_bound)]

fn bad_foo<T: Clone + Clone + Clone + Copy, U: Clone + Copy>(arg0: T, argo1: U) {
    unimplemented!();
}

fn bad_bar<T, U>(arg0: T, arg1: U)
where
    T: Clone + Clone + Clone + Copy,
    U: Clone + Copy,
{
    unimplemented!();
}

fn good_bar<T: Clone + Copy, U: Clone + Copy>(arg0: T, arg1: U) {
    unimplemented!();
}

fn good_foo<T, U>(arg0: T, arg1: U)
where
    T: Clone + Copy,
    U: Clone + Copy,
{
    unimplemented!();
}

trait GoodSelfTraitBound: Clone + Copy {
    fn f();
}

trait GoodSelfWhereClause {
    fn f()
    where
        Self: Clone + Copy;
}

trait BadSelfTraitBound: Clone + Clone + Clone {
    fn f();
}

trait BadSelfWhereClause {
    fn f()
    where
        Self: Clone + Clone + Clone;
}

trait GoodTraitBound<T: Clone + Copy, U: Clone + Copy> {
    fn f();
}

trait GoodWhereClause<T, U> {
    fn f()
    where
        T: Clone + Copy,
        U: Clone + Copy;
}

trait BadTraitBound<T: Clone + Clone + Clone + Copy, U: Clone + Copy> {
    fn f();
}

trait BadWhereClause<T, U> {
    fn f()
    where
        T: Clone + Clone + Clone + Copy,
        U: Clone + Copy;
}

struct GoodStructBound<T: Clone + Copy, U: Clone + Copy> {
    t: T,
    u: U,
}

impl<T: Clone + Copy, U: Clone + Copy> GoodTraitBound<T, U> for GoodStructBound<T, U> {
    // this should not warn
    fn f() {}
}

struct GoodStructWhereClause;

impl<T, U> GoodTraitBound<T, U> for GoodStructWhereClause
where
    T: Clone + Copy,
    U: Clone + Copy,
{
    // this should not warn
    fn f() {}
}

fn no_error_separate_arg_bounds(program: impl AsRef<()>, dir: impl AsRef<()>, args: &[impl AsRef<()>]) {}

fn main() {}
