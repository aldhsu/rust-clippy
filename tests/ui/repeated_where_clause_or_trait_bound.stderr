error: these bounds contain repeated elements
  --> $DIR/repeated_where_clause_or_trait_bound.rs:6:15
   |
LL | fn bad_foo<T: Clone + Clone + Clone + Copy, U: Clone + Copy>(arg0: T, argo1: U) {
   |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Clone + Copy`
   |
note: the lint level is defined here
  --> $DIR/repeated_where_clause_or_trait_bound.rs:4:9
   |
LL | #![deny(clippy::repeated_where_clause_or_trait_bound)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: these where clauses contain repeated elements
  --> $DIR/repeated_where_clause_or_trait_bound.rs:12:8
   |
LL |     T: Clone + Clone + Clone + Copy,
   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Clone + Copy`

error: these bounds contain repeated elements
  --> $DIR/repeated_where_clause_or_trait_bound.rs:40:26
   |
LL | trait BadSelfTraitBound: Clone + Clone + Clone {
   |                          ^^^^^^^^^^^^^^^^^^^^^ help: try: `Clone`

error: these where clauses contain repeated elements
  --> $DIR/repeated_where_clause_or_trait_bound.rs:47:15
   |
LL |         Self: Clone + Clone + Clone;
   |               ^^^^^^^^^^^^^^^^^^^^^ help: try: `Clone`

error: these bounds contain repeated elements
  --> $DIR/repeated_where_clause_or_trait_bound.rs:61:24
   |
LL | trait BadTraitBound<T: Clone + Clone + Clone + Copy, U: Clone + Copy> {
   |                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Clone + Copy`

error: these where clauses contain repeated elements
  --> $DIR/repeated_where_clause_or_trait_bound.rs:68:12
   |
LL |         T: Clone + Clone + Clone + Copy,
   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Clone + Copy`

error: these bounds contain repeated elements
  --> $DIR/repeated_where_clause_or_trait_bound.rs:101:19
   |
LL | fn bad_generic<T: GenericTrait<u64> + GenericTrait<u32> + GenericTrait<u64>>(arg0: T) {
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `GenericTrait<u32> + GenericTrait<u64>`

error: these bounds contain repeated elements
  --> $DIR/repeated_where_clause_or_trait_bound.rs:109:22
   |
LL | fn qualified_path<T: std::clone::Clone + Clone + foo::Clone>(arg0: T) {
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Clone + foo::Clone`

error: aborting due to 8 previous errors

